static/js/choropleth/choropleth.ts

```diff
@@ -0,0 +1,178 @@
+        .on("mouseover", handleMapHover)
+        .on("mouseleave", mouseLeave)
+        .on("click", handleMapClick);
+}
+
+/**
+ * Capture hover event on country and displays relevant information.
+ * @param {json} country is the geoJson content for the hovered country.
+ */
+function handleMapHover(country) {
+    // Display statistical variable information on hover.
+    let name = country.properties.name;
+    if (country.properties.hasOwnProperty('value')) {
+        document.getElementById("display").innerHTML = name + " - " +
+            country.properties.value;
+    } else {
+        document.getElementById("display").innerHTML = name + " - No Value";
+    }
+    
+    // Highlight selected country in black on hover.
+    d3.select(this).style("stroke", "black");
+  }
+  
+  /**
+   * Clears output after leaving a country.
+   */
+  function mouseLeave() {
+    // Remove hover text.
+    document.getElementById("display").innerHTML = "";
+  
+    // Remove country display effect.
+    d3.select(this).style("stroke", "white");
+  }
+  
+  /**
+   * Capture click event on country and redirects
+   * user to that page in the browser.
+   * @param {json} country is the geoJson content for the clicked country.
+   */
+  function handleMapClick(geo) {
+    if (geo.properties.hasSublevel) {
+      redirectToGeo(geo.properties.geoId)
+    } else {
+      alert("This geo has no further sublevels!")
+    }
+  }
+
+  function redirectToGeo(geoId) {
+    var base_url = document.location.origin + document.location.pathname + '?'
+    base_url += "statVar=" + url.searchParams.get("statVar");
+    base_url += "&perCapita=" + url.searchParams.get("perCapita")
+    base_url += "&place=" + geoId
+    base_url += "&bc=" 
+    
+    // Add or create breadcrumbs field.
+    var breadcrumbs = url.searchParams.get("bc");
+    if (breadcrumbs != null && breadcrumbs != "") {
+      base_url += breadcrumbs + ";"
+    }
+    base_url += url.searchParams.get("place");
+    window.location.href = base_url
+  }
+
+  /**
+   * Generates the breadcrumbs text from browswer url.
+   */
+  function generateBreadCrumbs() {
+    var breadcrumbs = url.searchParams.get("bc");
+    var breadcrumbs_display = document.getElementById("breadcrumbs");
+    var crumbs = breadcrumbs.split(";")
+  
+    // Build url each reference in the breadcrumbs.
+    var base_url = document.location.origin + document.location.pathname + '?'
+    base_url += "statVar=" + url.searchParams.get("statVar");
+    base_url += "&perCapita=" + url.searchParams.get("perCapita")
+    base_url += "&place=";
+
+    var breadcrumbs_upto = "";
+    for (let index in crumbs) {
+      let level_ref = crumbs[index];
+  
+      if (level_ref != "") {
+        // TODO how should we sanitize this?
+        let curr_url = base_url + level_ref + "&bc=" + breadcrumbs_upto;
+        breadcrumbs_display.innerHTML += 
+          "<a href=\"" + curr_url + "\">" + level_ref +  "</a>" + " > ";
```

Comment:
Let's try avoid building elements this way (it's a security risk, for one). Ok for now, but will get updated as we use React for more components.

Reply:
I will add a todo